File: src/main/java/org/hisp/dhis/expression/eval/NodeInterpreter.java
package org.hisp.dhis.expression.eval;

import org.hisp.dhis.expression.ast.BinaryOperator;
import org.hisp.dhis.expression.ast.DataItemModifier;
import org.hisp.dhis.expression.ast.NamedFunction;
import org.hisp.dhis.expression.ast.NamedValue;
import org.hisp.dhis.expression.ast.Node;
import org.hisp.dhis.expression.ast.UnaryOperator;
import org.hisp.dhis.expression.spi.DataItemType;

import java.time.LocalDate;
import java.util.function.Function;

/**
 * A {@link NodeInterpreter} interprets a root {@link Node} to calculate or derive some value from it.
 *
 * @author Jan Bernitt
 *
 * @param <T> type of the result of the interpretation
 */
public interface NodeInterpreter<T> extends Function<Node<?>, T> {

    @Override
    @SuppressWarnings("unchecked")
    default T apply(Node<?> node) {
        switch(node.getType()) {
            // complex nodes
            case UNARY_OPERATOR: return evalUnaryOperator((Node<UnaryOperator>) node);
            case BINARY_OPERATOR: return evalBinaryOperator((Node<BinaryOperator>) node);
            case ARGUMENT: return evalArgument((Node<Integer>) node);
            case PAR: return evalParentheses((Node<Void>) node);
            case FUNCTION: return evalFunction((Node<NamedFunction>) node);
            case MODIFIER: return evalModifier((Node<DataItemModifier>) node);
            case DATA_ITEM: return evalDataItem((Node<DataItemType>) node);

            // simple nodes
            case BOOLEAN: return evalBoolean((Node<Boolean>) node);
            case UID: return evalUid((Node<String>) node);
            case DATE: return evalDate((Node<LocalDate>) node);
            case NULL: return evalNull((Node<Void>) node);
            case NUMBER: return evalNumber((Node<Double>) node);
            case STRING: return evalString((Node<String>) node);
            case INTEGER: return evalInteger((Node<Integer>) node);
            case IDENTIFIER: return evalIdentifier(node);
            case NAMED_VALUE: return evalNamedValue((Node<NamedValue>) node);

            default: throw new UnsupportedOperationException("Not type not supported yet: "+node.getType());
        }
    }

    default T evalParentheses(Node<Void> group) {
        return group.child(0).eval(this);
    }

    default T evalArgument(Node<Integer> argument) {
        return argument.child(0).eval(this);
    }

    T evalBinaryOperator(Node<BinaryOperator> operator) ;

    T evalUnaryOperator(Node<UnaryOperator> operator) ;

    T evalFunction(Node<NamedFunction> function) ;

    T evalModifier(Node<DataItemModifier> modifier) ;

    T evalDataItem(Node<DataItemType> data) ;

    /*
    Simple nodes:
     */

    T evalNamedValue(Node<NamedValue> value);

    T evalNumber(Node<Double> value) ;

    T evalInteger(Node<Integer> value);

    T evalBoolean(Node<Boolean> value);

    T evalNull(Node<Void> value);

    T evalString(Node<String> value);

    T evalIdentifier(Node<?> value);

    T evalUid(Node<String> value) ;

    T evalDate(Node<LocalDate> value);
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/eval/NodeInterpreter.kt
package org.hisp.dhis.lib.expression.eval

import org.hisp.dhis.lib.expression.ast.*
import org.hisp.dhis.lib.expression.spi.DataItemType
import java.time.LocalDate

/**
 * A [NodeInterpreter] interprets a root [Node] to calculate or derive some value from it.
 *
 * @param <T> type of the result of the interpretation
 * @author Jan Bernitt
</T> */
interface NodeInterpreter<T> : (Node<*>) -> T? {

    @Suppress("kotlin:S6530", "UNCHECKED_CAST")
    override fun invoke(node: Node<*>): T? {
        return when (node.getType()) {
            NodeType.UNARY_OPERATOR -> evalUnaryOperator(node as Node<UnaryOperator>)
            NodeType.BINARY_OPERATOR -> evalBinaryOperator(node as Node<BinaryOperator>)
            NodeType.ARGUMENT -> evalArgument(node as Node<Int>)
            NodeType.PAR -> evalParentheses(node as Node<Unit>)
            NodeType.FUNCTION -> evalFunction(node as Node<NamedFunction>)
            NodeType.MODIFIER -> evalModifier(node as Node<DataItemModifier>)
            NodeType.DATA_ITEM -> evalDataItem(node as Node<DataItemType>)
            NodeType.VARIABLE -> evalVariable(node as Node<VariableType>)
            NodeType.BOOLEAN -> evalBoolean(node as Node<Boolean>)
            NodeType.UID -> evalUid(node as Node<String>)
            NodeType.DATE -> evalDate(node as Node<LocalDate>)
            NodeType.NULL -> evalNull(node as Node<Unit>)
            NodeType.NUMBER -> evalNumber(node as Node<Double>)
            NodeType.STRING -> evalString(node as Node<String>)
            NodeType.INTEGER -> evalInteger(node as Node<Int>)
            NodeType.IDENTIFIER -> evalIdentifier(node as Node<Any>)
            NodeType.NAMED_VALUE -> evalNamedValue(node as Node<NamedValue>)
        }
    }

    fun evalParentheses(group: Node<Unit>): T? {
        return group.child(0).eval(this)
    }

    fun evalArgument(argument: Node<Int>): T? {
        return argument.child(0).eval(this)
    }

    fun evalBinaryOperator(operator: Node<BinaryOperator>): T?
    fun evalUnaryOperator(operator: Node<UnaryOperator>): T?
    fun evalFunction(fn: Node<NamedFunction>): T?
    fun evalModifier(modifier: Node<DataItemModifier>): T
    fun evalDataItem(item: Node<DataItemType>): T?
    fun evalVariable(variable: Node<VariableType>): T?

    /*
    Simple nodes:
     */
    fun evalNamedValue(value: Node<NamedValue>): T
    fun evalNumber(value: Node<Double>): T
    fun evalInteger(value: Node<Int>): T
    fun evalBoolean(value: Node<Boolean>): T
    fun evalNull(value: Node<Unit>): T?
    fun evalString(value: Node<String>): T
    fun evalIdentifier(value: Node<Any>): T
    fun evalUid(value: Node<String>): T
    fun evalDate(value: Node<LocalDate>): T
}
